### PR TITLE
celeste-classic: init at unstable-2020-12-08

### PR DESCRIPTION
**Pull request by mrtnvgr**
## Description of changes

This PR adds [celeste-classic](https://celesteclassic.github.io/) game with optional support for Practice Mod.

NOTE: requires [this](https://github.com/NixOS/nixpkgs/pull/258963/commits/b5f97b03a312d72bc19da535a504f25db481139e) commit

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  src = fetchzip {
+    url = "https://www.speedrun.com/static/resource/174ye.zip";
+    hash = "sha256-GANHqKB0N905QJOLaePKWkUuPl9UlL1iqvkMMvw/CC8=";
+  };
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+  ];
+
+  buildInputs = with pkgs; [ SDL2 ];
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:20]: move SDL2 to derivation inputs and remove `with pkgs;`

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  sourceRoot = ".";
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:22]: what's the purpose of `sourceRoot = ".";` here? Wouldn't removing `*/` from `install` commands along with this line be fine?

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    install -Dsm755 */${folder}/${srcbin} $out/bin/${outbin}
+    install -Dm444 */${folder}/data.pod $out/bin/data.pod
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "A PICO-8 platformer about climbing a mountain, made in four days ${if practiceMod then "(Practice Mod)" else ""}";
+    homepage = "https://celesteclassic.github.io/";
+    license = licenses.unfree;
+    platforms = platforms.linux;
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:35]:
```suggestion
    platforms = platforms.linux;
    mainProgram = outbin;
```

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
+{ stdenvNoCC, fetchzip, autoPatchelfHook, pkgs, lib, practiceMod ? false }:
+
+let
+  folder = if practiceMod then "CELESTE*Practice*" else "CELESTE";
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:4]: small nitpick: Linux people usually prefer "directory" to "folder" (you can rename the var to `dir`)

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 5]
```diff
@@ -0,0 +1,39 @@
+{ stdenvNoCC, fetchzip, autoPatchelfHook, pkgs, lib, practiceMod ? false }:
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:1]:
```suggestion
{ lib
, stdenvNoCC
, fetchzip
, autoPatchelfHook
, SDL2
, practiceMod ? false
}:
```

**Review by chayleaf — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ce/celeste-classic/package.nix (hunk 2 above)
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:22]: there's only a single top-level directory in the zip, so I don't think this is needed? 
```suggestion
```

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  ];
+
+  buildInputs = with pkgs; [ SDL2 ];
+
+  sourceRoot = ".";
+
+  installPhase = ''
+    runHook preInstall
+    install -Dsm755 */${folder}/${srcbin} $out/bin/${outbin}
+    install -Dm444 */${folder}/data.pod $out/bin/data.pod
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:27]: If you remove `sourceRoot = ".";`, this is also necessary

```patch
-    install -Dsm755 */${folder}/${srcbin} $out/bin/${outbin}
-    install -Dm444 */${folder}/data.pod $out/bin/data.pod
+    install -Dsm755 ${folder}/${srcbin} $out/bin/${outbin}
+    install -Dm444 ${folder}/data.pod $out/bin/data.pod
```
```

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+  installPhase = ''
+    runHook preInstall
+    install -Dsm755 ${directory}/${srcbin} $out/bin/${outbin}
+    install -Dm444 ${directory}/data.pod $out/bin/data.pod
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "A PICO-8 platformer about climbing a mountain, made in four days ${if practiceMod then "(Practice Mod)" else ""}";
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:36]:
```suggestion
    description = "A PICO-8 platformer about climbing a mountain, made in four days${if practiceMod then " (Practice Mod)" else ""}";
```

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+, autoPatchelfHook
+, SDL2
+, practiceMod ? false
+}:
+
+let
+  directory = if practiceMod then "CELESTE*Practice*" else "CELESTE";
+  srcbin = if practiceMod then "celeste_practice_mod" else "celeste";
+  outbin = if practiceMod then "celeste-classic-pm" else "celeste-classic";
+in stdenvNoCC.mkDerivation {
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:13]: imo, breaking the line here would look less cluttered
```suggestion
in
stdenvNoCC.mkDerivation {
```

**Comment by chayleaf** [comment]
please squash this, the diff LGTM otherwise (will approve later when I check the binaries)

**Comment by chayleaf** [comment]
by the way, you can put the maintainer commit in this PR as well, git is smart enough to not apply it twice

**Comment by chayleaf** [comment]
the order has to be reverse (first the maintainer commit, then the package)

**Comment by mrtnvgr** [comment]
> the order has to be reverse (first the maintainer commit, then the package; this is what the rebase command i sent would've done)

Thanks a lot! Is it normal that this tries to rebase 10566 commits?

**Comment by chayleaf** [comment]
sorry for the confusion, in the current state you can do `git rebase -i HEAD~2` and reorder the commits (just swap the two lines)

**Comment by chayleaf** [comment]
Result of `nixpkgs-review pr 264717` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>celeste-classic</li>
  </ul>
</details>

**Comment by chayleaf** [comment]
do you want to expose `celeste-classic-pm` in `pkgs` (also, not sure about the descriptivity of the name)? If so, add it to all-packages:

```nix
celeste-classic-pm = pkgs.celeste-classic.override {
  practiceMod = true;
};
```

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  nativeBuildInputs = [
+    autoPatchelfHook
+  ];
+
+  buildInputs = [ SDL2 ];
+
+  installPhase = ''
+    runHook preInstall
+    install -Dsm755 ${directory}/${srcbin} $out/bin/${outbin}
+    install -Dm444 ${directory}/data.pod $out/bin/data.pod
```
Comment (chayleaf) [inline pkgs/by-name/ce/celeste-classic/package.nix:32]: This puts `data.pod` to `bin`, while it isn't a binary. Instead, do something like this:

```patch
-    install -Dsm755 ${directory}/${srcbin} $out/bin/${outbin}
-    install -Dm444 ${directory}/data.pod $out/bin/data.pod
+    install -Dsm755 ${directory}/${srcbin} $out/lib/${outbin}/${outbin}
+    install -Dm444 ${directory}/data.pod $out/lib/${outbin}/data.pod
+    mkdir -p $out/bin
+    ln -s $out/lib/${outbin}/${outbin} $out/bin/
```

**Review by chayleaf — APPROVED** [review]
A better approach than this would be packaging [picolove](https://github.com/picolove/picolove) and making this simply run the celeste-classic cartridge via picolove (this would keep unfree lua code but remove unfree native code built by random people on speedrun.com). But it's completely up to you, and I'm approving this since it works fine.

**Comment by mrtnvgr** [comment]
> A better approach than this would be packaging [picolove](https://github.com/picolove/picolove) and making this simply run the celeste-classic cartridge via picolove (this would keep unfree lua code but remove unfree native code). But it's completely up to you, and I'm approving this since it works fine.

Yeah, I think this would be nice, but the repo looks unmaintained and its not packaged for nix yet. I'll try to make some wrapper function (e.g. `mkPico8`), but I'm not sure how and what name should I use?

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+  outbin = if practiceMod then "celeste-classic-pm" else "celeste-classic";
+in
+stdenvNoCC.mkDerivation {
+  pname = outbin;
+  version = "1.0";
+
+  src = fetchzip {
+    url = "https://www.speedrun.com/static/resource/174ye.zip";
+    hash = "sha256-GANHqKB0N905QJOLaePKWkUuPl9UlL1iqvkMMvw/CC8=";
+  };
```
Comment (ckiee) [inline pkgs/by-name/ce/celeste-classic/package.nix:21]: Pin it to what looks to a specific version (so this doesn't break or rely on the cache if there's ever a new version) and note where it came from.
```suggestion
	# From https://www.speedrun.com/celestep8/resources
  src = fetchzip {
    url = "https://www.speedrun.com/static/resource/174ye.zip?v=f3dc98f";
    hash = "sha256-GANHqKB0N905QJOLaePKWkUuPl9UlL1iqvkMMvw/CC8=";
  };
```

**Comment by chayleaf** [comment]
no need for `mkPico8`, look at how love games are packaged (`pkgs/games/{90secondportraits,duckmarines,mari0,mrrescue,orthorobot,sienna,wireworld,stone-kingdoms}`)

File: pkgs/by-name/ce/celeste-classic/package.nix [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+}:
+
+let
+  directory = if practiceMod then "CELESTE*Practice*" else "CELESTE";
+  srcbin = if practiceMod then "celeste_practice_mod" else "celeste";
+  outbin = if practiceMod then "celeste-classic-pm" else "celeste-classic";
+in
+stdenvNoCC.mkDerivation {
+  pname = outbin;
+  version = "1.0";
```
Comment (ckiee) [inline pkgs/by-name/ce/celeste-classic/package.nix:16]: We are told this is:
```suggestion
  version = "unstable-2020-12-08";
```
…by hovering on the "2 years ago" text.

**Review by ckiee — CHANGES_REQUESTED** [review]
thank you! looks mostly good (:

**Review by ckiee — APPROVED** [review]
This looks happy to me. I'll merge if you want me to, then it's up to you to decide whether you want to open another PR for a source build in the future.

**Comment by ckiee** [comment]
Renamed the commit/PR and fixed a tiny nit, now using `optionalString` on the meta.description line.

**Comment by mrtnvgr** [comment]
> Renamed the commit/PR and fixed a tiny nit, now using `optionalString` on the meta.description line. The reaction means yes merge?

Yes, please

**Comment by ckiee** [comment]
`fetchzip` did not like the ?v= query string. Added `extension = "zip";`, building again.

**Comment by ckiee** [comment]
Result of `nixpkgs-review pr 264717` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>celeste-classic</li>
    <li>celeste-classic-pm</li>
  </ul>
</details>

+launches and runs fine

**Comment by ckiee** [comment]
congrats! you should get an invite to the nixos github org soon enough, welcome to nixpkgs :tada:

**Comment by mrtnvgr** [comment]
> congrats! you should get an invite to the nixos github org soon enough, welcome to nixpkgs 🎉

Thank you all! ♥️♥️♥️

**Pull request merged**